server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java

```diff
@@ -156,54 +167,164 @@ private void finishLeaf() {
+        Sort indexSort = context.reader().getMetaData().getSort();
+        if (indexSort == null) {
+            return null;
+        }
+        List<SortField> sortFields = new ArrayList<>();
+        for (int i = 0; i < indexSort.getSort().length; i++) {
+            SingleDimensionValuesSource<?> source = sources[i];
+            SortField indexSortField = indexSort.getSort()[i];
+            if (source.fieldType == null
+                    // TODO: can we handle missing bucket when using index sort optimization ?
+                    || source.missingBucket
+                    || indexSortField.getField().equals(source.fieldType.name()) == false
+                    || isMultiValued(context, indexSortField)) {
+                    // TODO: ignore fields that change the value with a script
+                break;
+            }
+
+            if (indexSortField.getReverse() != (source.reverseMul == -1)) {
+                if (i == 0) {
+                    // the leading index sort matches the leading source field but the order is reversed
+                    // so we don't check the other sources.
+                    return new Sort(indexSortField);
+                }
+                break;
+            }
+            sortFields.add(indexSortField);
+        }
+        return sortFields.isEmpty() ? null : new Sort(sortFields.toArray(new SortField[0]));
+    }
+
+    /**
+     * Return the number of leading sources that match the index sort.
+     *
+     * @param indexSortPrefix The index sort prefix that matches the sources
+     * @return The length of the index sort prefix if the sort order matches
+     *         or -1 if the leading index sort is in the reverse order of the
+     *         leading source. A value of 0 indicates that the index sort is
+     *         not applicable.
+     */
+    private int computeSortPrefixLen(Sort indexSortPrefix) {
+        if (indexSortPrefix == null) {
+            return 0;
+        }
+        if (indexSortPrefix.getSort()[0].getReverse() != (sources[0].reverseMul == -1)) {
+            assert indexSortPrefix.getSort().length == 1;
+            return -1;
+        } else {
+            return indexSortPrefix.getSort().length;
+        }
+    }
+
+    private void processLeafFromQuery(LeafReaderContext ctx, Sort indexSortPrefix) throws IOException {
+        DocValueFormat[] formats = new DocValueFormat[indexSortPrefix.getSort().length];
+        for (int i = 0; i < formats.length; i++) {
+            formats[i] = sources[i].format;
+        }
+        FieldDoc fieldDoc = SearchAfterBuilder.buildFieldDoc(new SortAndFormats(indexSortPrefix, formats),
+            Arrays.copyOfRange(rawAfterKey.values(), 0, formats.length));
+        if (indexSortPrefix.getSort().length < sources.length) {
+            // include all docs that belong to the partial bucket
+            fieldDoc.doc = 0;
+        }
+        BooleanQuery newQuery = new BooleanQuery.Builder()
```

Comment:
> Here would use original context.query() + search after query to skip mismatched docs, I am considering the performance issue, since context.query() should also be executed on the top level of the search request. That means context.query() would be executed twice.

This is why we need to update the documentation ;). The idea is that the query and the aggregation are independent and have different early termination logic. If you don't need the top hits and the number of hits that match the query you can execute the request with `"size": 0` and `"track_total_hits": false`. This will shortcut the query part to not run the query unless an aggregation is provided. If the provided aggregation is a `composite`, the early termination logic will be propagated such as the entire collection will finish.

> In my PR's logic, I use start doc id (got from after key value by inverted lists or bkd-tree) to filter out the mismatched docs, this would avoid twice queries. Here is the related code block in CompositeAggregator.java of my PR:
if (doc >= startDocId) { inner.collect(doc); }
How do you think these two approaches?

The approach in this pr is to directly skip the documents from the previous pages rather than filtering them one by one with the start doc id. I think it can make a big difference if you need to filter a lot of documents from the previous pages since modifying the query allows to jump directly to the documents that can be competitive.

> Also shall we need to add some tests to verify this mismatched docs filtering optimization?

I modified the testCase in the CompositeAggregatorTests to run with and without index sort. I also added specific tests in the CompositeValuesCollectorQueueTests to check early termination. Do you have a specific test in mind that is missing ?

Reply:
> This is why we need to update the documentation ;). The idea is that the query and the aggregation are independent and have different early termination logic. If you don't need the top hits and the number of hits that match the query you can execute the request with `"size": 0` and `"track_total_hits": false`. This will shortcut the query part to not run the query unless an aggregation is provided. If the provided aggregation is a `composite`, the early termination logic will be propagated such as the entire collection will finish.

Got the point, we do need to update the document to tell users to shortcut the query part if only focus on the aggregation. :)

Reply:
> I modified the testCase in the CompositeAggregatorTests to run with and without index sort. I also added specific tests in the CompositeValuesCollectorQueueTests to check early termination. Do you have a specific test in mind that is missing ?

I was considering that whether we need to make sure  `CompositeAggregator.java#processLeafFromQuery` could be triggered and we do skipped documents directly, probably enough already.